src/main/java/org/sagebionetworks/bridge/services/AdherenceService.java

```diff
@@ -94,35 +98,95 @@ public void updateAdherenceRecords(String appId, AdherenceRecordList recordList)
                 AuthEvaluatorField.STUDY_ID, recordList.getRecords().get(0).getStudyId(), 
                 AuthEvaluatorField.USER_ID, recordList.getRecords().get(0).getUserId());
 
-        dao.updateAdherenceRecords(recordList);
-
         for (AdherenceRecord record : recordList.getRecords()) {
-            if (record.getFinishedOn() == null) {
-                continue;
+            dao.updateAdherenceRecord(record);
+        }
+        for (AdherenceRecord record : recordList.getRecords()) {
+            updateSessionState(appId, record);
+        }
+    }
```

Comment:
Prevent multiple updates of the same session event, which could create N publication updates in the history.

Reply:
Keep track of and do not duplicate session publication events.

Reply:
If every assessment has been declined, decline the session.